lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,477 @@
+        /// Driver tries to close all the running tasks and clean the running task collection in the end.
+        /// If all the tasks are running, the total number of running tasks should be _totalExpectedTasks -1
+        /// If this happens before all the tasks are running, then the total number of running tasks should smaller than _totalExpectedTasks -1
+        /// If this happens when no task is running, the total number of running tasks could be 0
+        /// </summary>
+        internal void CloseAllRunningTasks(string closeMessage)
+        {
+            Logger.Log(Level.Verbose, "Closing [{0}] running tasks.", _runningTasks.Count);
+            foreach (var runningTask in _runningTasks.Values)
+            {
+                runningTask.Dispose(Encoding.UTF8.GetBytes(closeMessage));
+                UpdateState(runningTask.Id, TaskStateEvent.WaitingTaskToClose);
+            }
+            _runningTasks.Clear();
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IRunningTask event but system is either in shutting down or fail.
+        /// In this case, the task should not be added in Running Tasks yet.
+        /// Change the task state to TaskRunning if it is still in TaskSubmitted state
+        /// Closes the IRunningTask 
+        /// Then move the task state to WaitingTaskToClose
+        /// Throw IMRUSystemException if runningTask is null or the running task is already added in the running task collection
+        /// </summary>
+        /// <param name="runningTask"></param>
+        /// <param name="closeMessage"></param>
+        internal void RecordRunningTaskDuringSystemFailure(IRunningTask runningTask, string closeMessage)
+        {
+            if (runningTask == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("RunningTask is null."), Logger);
+            }
+
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] is already in running tasks.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (GetTaskState(runningTask.Id) == StateMachine.TaskState.TaskSubmitted)
+            {
+                UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+            }
+            runningTask.Dispose(Encoding.UTF8.GetBytes(closeMessage));
+            UpdateState(runningTask.Id, TaskStateEvent.WaitingTaskToClose);
+        }
+
+        /// <summary>
+        /// Gets error type based on the information in IFailedTask 
+        /// Currently we use the Message in IFailedTask to distinguish different types of errors
+        /// </summary>
+        /// <param name="failedTask"></param>
+        /// <returns></returns>
+        private TaskStateEvent GetTaskErrorEvent(IFailedTask failedTask)
```

Comment:
I am wondering if instead of getting these message strings, IFailedTask should also return an Exception variable which can then be compared to what kind of exception it is (instead of string) and then generate events?